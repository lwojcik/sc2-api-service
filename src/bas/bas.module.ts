import { HttpModule } from '@nestjs/axios';
import { Module } from '@nestjs/common';
import { ConfigModule } from '@nestjs/config';
import { LoggerModule } from '../logger/logger.module';
import { basConfig } from '../config/bas.config';
import { BasService } from './bas.service';

@Module({
  imports: [
    ConfigModule.forFeature(basConfig),
    HttpModule.register({
      timeout: 5000,
      maxRedirects: 5,
    }),
    LoggerModule,
  ],
  providers: [BasService],
  exports: [BasService],
})
export class BasModule {}
