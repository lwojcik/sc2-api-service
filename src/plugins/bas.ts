import { FastifyInstance } from 'fastify';
import fp from 'fastify-plugin';
import http from 'http';

export interface BasOptions {
  url: String;
  statusEndpoint: String;
  accessTokenEndpoint: String;
  accessTokenRefreshEndpoint: String;
}

interface BASReply {
  status: number;
  data: {
    accessToken: string;
  };
}

interface OKReply {
  status: 200;
  message: 'ok';
}

const bas = fp(
  async (fastify: FastifyInstance, opts: BasOptions, next: Function) => {
    let isUp = false;
    const statusUrl = `${opts.url}/${opts.statusEndpoint}`;
    const accessTokenUrl = `${opts.url}/${opts.accessTokenEndpoint}`;
    const accessTokenRefreshUrl = `${opts.url}/${opts.accessTokenRefreshEndpoint}`;

    const basDown = {
      status: 500,
      message: 'Bnet-auth-service is down!',
    };

    /* istanbul ignore next */
    const get = (url: string): object =>
      new Promise((resolve, reject) => {
        http
          .get(url, (res) => {
            res.setEncoding('utf8');
            let body = '';
            res.on('data', chunk => (body += chunk));
            res.on('end', () => resolve(JSON.parse(body)));
          })
          .on('error', reject);
      });

    /* istanbul ignore next */
    const checkIfHostIsUp = async (url: string) => {
      try {
        const response = (await get(url)) as OKReply;
        if (response.status && response.status !== 200) return false;
        return true;
      } catch (error) {
        return false;
      }
    };

    /* istanbul ignore next */
    const checkOnStartup = async () => {
      const isBASup = await checkIfHostIsUp(statusUrl);
      isUp = isBASup;
      isBASup
        ? fastify.log.info('Bnet-auth-service status: running')
        : fastify.log.error('Bnet-auth-service is down!');
    };

    /* istanbul ignore next */
    const getAccessToken = async (refresh?: boolean) => {
      if (!isUp) return basDown;
      const basData = !refresh
        ? ((await get(accessTokenUrl)) as BASReply)
        : ((await get(accessTokenRefreshUrl)) as BASReply);
      const accessToken = basData.data.accessToken;
      return accessToken;
    };

    /* istanbul ignore next */
    const refreshAccessToken = () => getAccessToken(true);

    fastify.decorate('bas', {
      getAccessToken,
      refreshAccessToken,
    });

    checkOnStartup();

    next();
  },
);

export default bas;
