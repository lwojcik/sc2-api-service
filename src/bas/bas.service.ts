import { HttpService } from '@nestjs/axios';
import { Inject, Injectable } from '@nestjs/common';
import { ConfigType } from '@nestjs/config';
import { basConfig } from '../config/bas.config';
import { LoggerService } from '../logger/logger.service';

@Injectable()
export class BasService {
  constructor(
    @Inject(basConfig.KEY)
    private readonly basConf: ConfigType<typeof basConfig>,
    private readonly httpService: HttpService,
    private readonly logger: LoggerService
  ) {
    this.logger.setLoggedClass(BasService.name);
  }

  async getAccessToken() {
    const accessToken = await this.httpService.axiosRef.get<{
      accessToken: string;
    }>(`${this.basConf.url}/accesstoken`);
    return accessToken.data.accessToken;
  }
}
