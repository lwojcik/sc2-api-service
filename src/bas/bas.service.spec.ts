import { Test, TestingModule } from '@nestjs/testing';
import { BasService } from './bas.service';

describe('BasService', () => {
  let service: BasService;

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      providers: [BasService],
    }).compile();

    service = module.get<BasService>(BasService);
  });

  it('should be defined', () => {
    expect(service).toBeDefined();
  });
});
