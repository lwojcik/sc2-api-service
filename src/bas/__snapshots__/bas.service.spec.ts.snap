// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BasService should get access token 1`] = `"sample_access_token_from_httpservice_mock"`;

exports[`BasService should get refreshed access token 1`] = `"sample_access_token_from_httpservice_mock"`;
